include/revng/Pipeline/Containers.h

```diff
@@ -0,0 +1,227 @@
+    if (not llvm::sys::fs::exists(File)) {
+      Container = std::nullopt;
+      return llvm::Error::success();
+    }
+
+    return getOrCreate().loadFromDisk(File);
+  }
+
+  TargetsList enumerate() const {
+    if (not hasValue())
+      return {};
+
+    return get().enumerate();
+  }
+};
+
+/// a map from containerName to a MaybeContainer, which is initially empty.
+class Containers {
+public:
+  using Map = llvm::StringMap<MaybeContainer>;
+  using const_iterator = Map::const_iterator;
+
+private:
+  Map Content;
+
+public:
+  Containers(Map Content) : Content(std::move(Content)) {}
+  Containers() = default;
+  Containers(Containers &&) = default;
+  Containers &operator=(Containers &&) = default;
+  ~Containers() = default;
+  Containers(const Containers &) = delete;
+  Containers &operator=(const Containers &) = delete;
+
+  Containers cloneFiltered(const ContainersStatus &Targets);
+
+  void mergeContainers(Containers &&Other) {
+    for (auto &Entry : Other.Content) {
+      revng_assert(Content.count(Entry.first()) != 0);
+      MaybeContainer &LContainer = Content.find(Entry.first())->second;
+      MaybeContainer &RContainer = Entry.second;
+      LContainer.mergeBack(std::move(RContainer));
+    }
+  }
+
+  template<typename ContainerDerivedType>
+  ContainerDerivedType &getOrCreate(llvm::StringRef Name) {
+    revng_assert(containsOrCanCreate(Name));
+    auto &Base = Content.find(Name)->second.getOrCreate();
+    revng_assert(llvm::isa<ContainerDerivedType>(Base));
+    return llvm::cast<ContainerDerivedType>(Base);
+  }
+
+  template<typename ContainerDerivedType>
+  const ContainerDerivedType &get(llvm::StringRef Name) const {
+    revng_assert(contains(Name));
+    const auto &Base = Content.find(Name)->second.get();
+    revng_assert(llvm::isa<ContainerDerivedType>(Base));
+    return llvm::cast<ContainerDerivedType>(Base);
+  }
+
+  ContainerBase &get(llvm::StringRef Name) {
+    revng_assert(contains(Name));
+    return Content.find(Name)->second.getOrCreate();
+  }
+
+  const ContainerBase &get(llvm::StringRef Name) const {
+    revng_assert(contains(Name));
+    return Content.find(Name)->second.get();
+  }
+
+  bool contains(llvm::StringRef Name) const {
+    return Content.count(Name) != 0 and Content.find(Name)->second.hasValue();
+  }
+
+  bool containsOrCanCreate(llvm::StringRef Name) const {
+    return Content.count(Name) != 0;
+  }
+
+  bool contains(const Target &Target) const;
+
+  void addEmpty(llvm::StringRef Name, ContainerFactory &Factory) {
+    Content.insert_or_assign(Name, MaybeContainer(Factory));
+  }
+
+  void add(llvm::StringRef Name, MaybeContainer Container) {
+    Content.insert_or_assign(Name, std::move(Container));
+  }
+
+  void addNonEmpty(llvm::StringRef Name,
+                   ContainerFactory &Factory,
+                   std::unique_ptr<ContainerBase> Container) {
+    Content.insert_or_assign(Name,
+                             MaybeContainer(Factory, std::move(Container)));
+  }
```

Comment:
These methods confuse me, can't we a simple `add` method and let users do their thing?
What does it mean to add to a set of containers a factory? You *register* factories usually, is this the place for that?

Reply:
not sure what this would achieve? they are just helpers to avoid having to get the container after having added it to put stuff inside.

Reply:
You seem to like having these helpers, but they end up creating a quite large API that's not obvious to use to me.
I'd go for small, predicable and familiar APIs that are obvious to use, even if a bit more verbose on the user side.

If you want to have such a large API, you should thoroughly document it, otherwise I've to look at the implementation and, still, it's not obvious what's going on.